File: sections/68_tree_decompositions/problems/b_toggle_nearest_beacon/solution.py
import heapq, sys
from array import array


def main():
    tokens = sys.stdin.buffer.read().split()
    if not tokens:
        return
    idx = 0
    n = int(tokens[idx])
    q = int(tokens[idx + 1])
    idx += 2
    graph = [[] for _ in range(n)]
    for _ in range(n - 1):
        u = int(tokens[idx]) - 1
        v = int(tokens[idx + 1]) - 1
        idx += 2
        graph[u].append(v)
        graph[v].append(u)
    blocked = [False] * n
    paths = [array("i") for _ in range(n)]
    subtree = [0] * n
    todo = [0]
    while todo:
        entry = todo.pop()
        nodes = [entry]
        parent = {entry: -1}
        for u in nodes:
            for v in graph[u]:
                if not blocked[v] and v not in parent:
                    parent[v] = u
                    nodes.append(v)
        for u in reversed(nodes):
            subtree[u] = 1 + sum(
                subtree[v]
                for v in graph[u]
                if not blocked[v] and parent.get(v) != None and parent.get(v) == u
            )
        total = len(nodes)
        centroid = entry
        for u in nodes:
            largest = total - subtree[u]
            for v in graph[u]:
                if not blocked[v] and parent.get(v) == u:
                    largest = max(largest, subtree[v])
            if largest * 2 <= total:
                centroid = u
                break
        stack = [(centroid, -1, 0)]
        while stack:
            u, p, d = stack.pop()
            paths[u].extend((centroid, d))
            for v in graph[u]:
                if v != p and not blocked[v]:
                    stack.append((v, u, d + 1))
        blocked[centroid] = True
        for v in graph[centroid]:
            if not blocked[v]:
                todo.append(v)
    added = [[] for _ in range(n)]
    removed = [[] for _ in range(n)]
    active = [False] * n

    def clean(c):
        while added[c] and removed[c] and added[c][0] == removed[c][0]:
            heapq.heappop(added[c])
            heapq.heappop(removed[c])

    out = []
    for _ in range(q):
        kind = tokens[idx].decode()
        u = int(tokens[idx + 1]) - 1
        idx += 2
        if kind == "T":
            active[u] = not active[u]
            path = paths[u]
            for i in range(0, len(path), 2):
                c, d = path[i], path[i + 1]
                heapq.heappush(added[c] if active[u] else removed[c], d)
                if not active[u]:
                    clean(c)
        else:
            answer = 10**18
            path = paths[u]
            for i in range(0, len(path), 2):
                c, d = path[i], path[i + 1]
                clean(c)
                answer = min(answer, d + added[c][0]) if added[c] else answer
            out.append(str(-1 if answer == 10**18 else answer))
    print("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/68_tree_decompositions/problems/b_toggle_nearest_beacon/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q)) return 0;
  vector<vector<int>> g(n);
  for (int i = 1, u, v; i < n; i++) {
    cin >> u >> v;
    --u;
    --v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  vector<char> blocked(n), active(n);
  vector<int> parent(n, -1), subtree(n);
  vector<int> traversal_parent(n), seen(n);
  int stamp = 0;
  vector<vector<pair<int, int>>> path(n);
  function<void(int, int)> decompose = [&](int entry, int centroid_parent) {
    ++stamp;
    vector<int> nodes;
    nodes.push_back(entry);
    seen[entry] = stamp;
    traversal_parent[entry] = -1;
    for (int i = 0; i < (int)nodes.size(); i++) {
      int u = nodes[i];
      for (int v : g[u])
        if (!blocked[v] && seen[v] != stamp) {
          seen[v] = stamp;
          traversal_parent[v] = u;
          nodes.push_back(v);
        }
    }
    for (int i = nodes.size() - 1; i >= 0; i--) {
      int u = nodes[i];
      subtree[u] = 1;
      for (int v : g[u])
        if (!blocked[v] && seen[v] == stamp && traversal_parent[v] == u)
          subtree[u] += subtree[v];
    }
    int total = nodes.size(), centroid = entry;
    for (int u : nodes) {
      int largest = total - subtree[u];
      for (int v : g[u])
        if (!blocked[v] && seen[v] == stamp && traversal_parent[v] == u)
          largest = max(largest, subtree[v]);
      if (largest * 2 <= total) {
        centroid = u;
        break;
      }
    }
    parent[centroid] = centroid_parent;
    vector<tuple<int, int, int>> stack{{centroid, -1, 0}};
    while (!stack.empty()) {
      auto [u, p, d] = stack.back();
      stack.pop_back();
      path[u].push_back({centroid, d});
      for (int v : g[u])
        if (v != p && !blocked[v]) stack.push_back({v, u, d + 1});
    }
    blocked[centroid] = 1;
    for (int v : g[centroid])
      if (!blocked[v]) decompose(v, centroid);
  };
  decompose(0, -1);
  vector<multiset<int>> distances(n);
  while (q--) {
    char type;
    int u;
    cin >> type >> u;
    --u;
    if (type == 'T') {
      active[u] ^= 1;
      for (auto [c, d] : path[u]) {
        if (active[u])
          distances[c].insert(d);
        else
          distances[c].erase(distances[c].find(d));
      }
    } else {
      int answer = INT_MAX;
      for (auto [c, d] : path[u])
        if (!distances[c].empty()) answer = min(answer, d + *distances[c].begin());
      cout << (answer == INT_MAX ? -1 : answer) << '\n';
    }
  }
}
